File: 10_first_non_repeating.py
"""
Task 10: Find the first non-repeating character in a string.
Reads a line and prints the character or "_" if none.
"""
import sys
from collections import Counter

def first_non_repeating(s: str) -> str:
    cnt = Counter(s)
    for ch in s:
        if cnt[ch] == 1:
            return ch
    return '_'

if __name__ == '__main__':
    s = sys.stdin.readline().rstrip('\n')
    print(first_non_repeating(s))

# Time Complexity: O(n)
# Space Complexity: O(k)

File: 10_first_non_repeating.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 10: Find the first non-repeating character in a string.
Reads a line and prints the character or "_" if none.
*/

int main()
{
    string s;
    if (!getline(cin, s))
        return 0;
    vector<int> freq(256, 0);
    for (char c : s) ++freq[(unsigned char)c];
    for (char c : s)
        if (freq[(unsigned char)c] == 1)
        {
            cout << c << '\n';
            return 0;
        }
    cout << "_\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
